analyzers/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataFactory.cs

```diff
@@ -68,65 +66,37 @@ public static partial class NuGetMetadataFactory
 
         public static IEnumerable<MetadataReference> CreateNetStandard21()
         {
-            var packageDir = Path.Combine(PackagesFolder, @"NETStandard.Library.Ref.2.1.0\ref\netstandard2.1");
-            if (Directory.Exists(packageDir))
-            {
-                LogMessage($"Package found at {packageDir}");
-            }
-            else
-            {
-                LogMessage($"Package not found at {packageDir}");
-                PackageManager.InstallPackage("NETStandard.Library.Ref", SemanticVersion.ParseOptionalVersion("2.1.0"), ignoreDependencies: true, allowPrereleaseVersions: false);
-                if (!Directory.Exists(packageDir))
-                {
-                    throw new ApplicationException($"Test setup error: folder for downloaded package does not exist. Folder: {packageDir}");
-                }
-            }
+            var package = new Package("NETStandard.Library.Ref", "2.1.0", "netstandard2.1");
+            package.EnsurePackageIsInstalled();
 
-            return Directory.GetFiles(packageDir, "*.dll", SearchOption.AllDirectories)
+            return Directory.GetFiles(package.PackageDirectory(), "*.dll", SearchOption.AllDirectories)
                .Select(x => (MetadataReference)MetadataReference.CreateFromFile(x))
                .ToImmutableArray();
         }
 
         /// <param name="allowedDirectories">List of allowed directories sorted by preference to search for DLL files.</param>
         private static IEnumerable<MetadataReference> Create(Package package, string[] allowedDirectories)
         {
-            Console.WriteLine();
-            Console.WriteLine($"Package: {package.Id}, {package.Version}");
             package.EnsurePackageIsInstalled();
 
+            var allowedNugetLibDirectoriesByPreference = allowedDirectories.Select((folder, priority) => new { folder, priority });
             var packageDirectory = package.PackageDirectory();
-            if (!Directory.Exists(packageDirectory))
-            {
-                throw new ApplicationException($"Test setup error: Package directory doesn't exist: {packageDirectory}");
-            }
-            var dllsPerDirectory = Directory.GetFiles(packageDirectory, "*.dll", SearchOption.AllDirectories)
-                .GroupBy(x => new FileInfo(x).Directory.Name)
-                .ToDictionary(x => x.Key.Split('+').First(), x => x.AsEnumerable());
+            var matchingDllsGroups = Directory.GetFiles(packageDirectory, "*.dll", SearchOption.AllDirectories)
+                                              .Select(path => new FileInfo(path))
+                                              .GroupBy(file => file.Directory.Name).ToArray();
+            var selectedGroup = matchingDllsGroups.Length == 1 && matchingDllsGroups[0].Key.EndsWith(".dll")
+                ? matchingDllsGroups[0]
+                : matchingDllsGroups.Join(
```

Comment:
This painful `join` with weird grouping technique was very hard to understand and I refactored it to that more straight-forward way in this method. Can you restore that?

sonar-security is behind in this logic.